### ISSUE TITLE
Dart VM crash

### ISSUE THREAD

**Issue by alsemenov** [dart-lang/sdk#24935]
The following code lead to Dart VM crash
```
import 'dart:async';

StreamController<String> newStreamControllerString(int id, List<int> log){
//  print("newStreamControllerString $id \n");
  StreamController<String> sc = new StreamController<String>();
  sc.add('a'+id.toString());
  sc.add('b'+id.toString());
  sc.onCancel = () {
    log.add(id);
    print("cancelled: $id");
  };
  return sc;
}

void isFalse(var actual, [String reason = null]) {
if (_identical(actual, false)) return;
String msg = _getMessage(reason);
_fail("Expect.isFalse($actual$msg) fails.");
}

String _getMessage(String reason)
=> (reason == null) ? "" : ", '$reason'";

void _fail(String message) {
  throw new ExpectException(message);
}

bool _identical(a, b) => identical(a, b);

class ExpectException implements Exception {
  String message;
  ExpectException([this.message]);
  String toString() => message;
}

Future test3() async {
  List<int> log = [];
  int i = 1;
  StreamController<StreamController<String>> sc3 =
  new StreamController<StreamController<String>>();
  sc3.add(newStreamControllerString(i++, log));
  sc3.add(newStreamControllerString(i++, log));
  sc3.onCancel = () { log.add(0); };
  sc3.close();

  await for (StreamController<String> sc in sc3.stream){
    print('$sc');
    L: await for (String s in sc.stream){
      await for (String s in newStreamControllerString(i++, log).stream){
        break L;
      }
    }
    isFalse(sc3.isClosed);
  }
  print('$log');
}

main() {
  print("test start");
  test3().then((v) => print("test result: $v"))
  .then((v) => print("test finish"));
}
```
I am sorry for such a long test, but I am not able to make it shorter.
The test output is:
>test start
>Instance of '_AsyncStreamController<String>'
>cancelled: 3
>cancelled: 1
And then process finishes with exit code -1073741819 (0xC0000005)

Dart VM version: 1.12.2 (Wed Oct 21 07:10:52 2015) on "windows_x64"

**Comment by lrhn** [comment]
Reproduces on Linux as well. Even with a debug build, it just dumps core.

**Comment by lrhn** [comment]
Smaller reproduction:
```
import 'dart:async';

Future main() async {
  L: await for (var s in new Stream.fromIterable([1])){
    await for (var s in new Stream.fromIterable([1])){
      break L;  // Break past two await-for's.
    }
  }
  throw 0;  // Any throw here.
}
```

**Comment by alsemenov** [comment]
The same crash is observed with continue statement
```dart
import 'dart:async';

Future main() async {
  L:
  for (var s in [1]){ 
    await for (var s in new Stream.fromIterable([1])){
      await for (var s in new Stream.fromIterable([1])){
        continue L;  // Break past two await-for's.
      }
    }
  }
  throw 0;  // Any throw here.
}
```

**Comment by mhausner** [comment]
Hmm, I cannot reproduce this on a current tree on OS X. The version containing the break statement produces:

```
dart ~/tmp/s.dart
Unhandled exception:
0
#0      main.<main_async_body> (file:///Users/hausner/tmp/s.dart:12:3)
#1      _RootZone.runUnary (dart:async/zone.dart:1137)
...
```

Which is what I would expect.

**Comment by mhausner** [comment]
I was able to reproduce the crash when jumping out of the loop with a continue statement, but not with a break statement.

Fix is here:  https://codereview.chromium.org/1661433003/

**Comment by mhausner** [comment]
Fixed with fef73d59ad177c6689cec2d1bb103714727df978

Thank you for the small repro case!
